docs/website/blog/2023-06-01-writing-devfile-from-scratch.mdx

```diff
@@ -0,0 +1,359 @@
+- `composite` - this command can be used to execute multiple existing commands sequentially or in parallel.
+
+
+
+Let's add commands to our Defile.
+
+
+```yaml
+schemaVersion: 2.2.0
+metadata:
+  name: my-backstage
+
+commands:
+# highlight-start
+  - id: yarn-install
+    exec:
+      commandLine: npx yarn install
+      component: nodejs
+      workingDir: ${PROJECT_SOURCE}
+      group:
+        kind: build
+        isDefault: true
+    
+  - id: run-dev
+    exec:
+      commandLine: npx yarn run dev
+      component: nodejs
+      workingDir: ${PROJECT_SOURCE}
+      group:
+        kind: run
+        isDefault: true
+# highlight-end
+
+components:
+  - name: nodejs
+    container:
+      image: registry.access.redhat.com/ubi9/nodejs-18:latest
+      sourceMapping: /projects
+      command: ['tail', '-f', '/dev/null']
+      memoryLimit: 2Gi
+      endpoints:
+        - name: frontend
+          targetPort: 3000
+          exposure: public
+        - name: backend
+          targetPort: 7007
+          exposure: public
+```
+
+
+We have added two commands `yarn-install` and `run-dev`. Let's use the first one to explain what each line means.
+- `commandLine: npx yarn install` - this defines that the command `npx yarn install` should be executed when Devfile command
+  `yarn-install` is executed.
+- `component: nodejs` - this defines in which `container` component the command defined in `commandLine` should be executed.
+- `workingDir: ${PROJECT_SOURCE}` - defines in what working directory the command will be executed.
+  Here we are using `${PROJECT_SOURCE}` variable. This variable will always point to the root directory with the source code.
+  This will be the same path as we used in `sourceMapping` in the `container` component
+- `group:` - defines to what group this command belongs to. There are `build`, `run`, `debug`, `test`, `deploy`.
+  ```yaml
+        kind: build
+        isDefault: true
+  ```
+  The previous block defines that this command belongs to `build` group and is the default command. 
+  Each group can have only one default command. When you run `odo dev`, odo automatically executes the default command in `build` group first,
+ followed by the default command in `run` group.
+ 
+ 
+Ideally, this would be all we need, and you could use this Devfile with odo.
+
+## Fixing issues
+If you try to use Devfile as we have it, you will see an error.
+The first problem is that the NodeJS image doesn't have `yarn` installed. 
+
+### Install `yaml` into the container
+To add yarn, we will leverage Devfile feature called `events`.
+Events allow us to define commands that should be executed on predefined events. 
+There are 3 events that you can use.
+- `preStart` - executed before the main container is started.
+- `postStart` - executed after the main container is started.
+- `preStop` - executed before the main container is stopped.
```

Comment:
I think "main container" sound a little confusing, makes me wonder how can I define a main container if there are multiple containers present in the Devfile?

I might be overthinking this one, so feel free to ignore this comment.

Reply:
This is a good point. I was not sure about this term as well, but I was not able to come up with a better definition :-(